core/src/main/java/io/undertow/client/http/HttpRequestConduit.java

```diff
@@ -277,8 +278,11 @@ private int processWrite(int state, final ByteBuffer userData) throws IOExceptio
                         if (! buffer.hasRemaining()) {
                             buffer.flip();
                             do {
-                                res = next.write(buffer);
+                                res = writeNext(buffer);
                                 if (res == 0) {
+                                    this.headerName = headerName;
+                                    this.valueIterator = valueIterator;
+                                    this.nameIterator = nameIterator;
                                     log.trace("Continuation");
                                     return STATE_HDR_EOL_CR;
```

Comment:
The transitions to `STATE_HDR_EOL_LF` and `STATE_HDR_EOL_CR` require additional data.

I might suggest saving all state prior to every transition to avoid this kind of bug in the future, however it would require more operations per invocation. That said, in the common case I'd expect the request line and all headers to fit into a single buffer, so it's likely there's no real cost.